audio/src/conn.rs

```diff
@@ -1,86 +1,545 @@
-            self.state.time = time;
+    /// Start to play music if music isn't playing. Stop music if music is playing.
+    pub fn set_music(&mut self, state: &State) {
+        let play_state = *self.play_state.lock();
+        match play_state {
+            PlayState::NotPlaying => self.start_music(state),
+            _ => self.stop_music(&state.music),
         }
-        self.sample = match self.recv_sample.try_recv() {
-            Ok(sample) => Some(sample),
-            Err(_) => None,
-        };
-        // Get the export state.
-        if self.export_state.is_some() {
-            self.send(vec![Command::SendExportState]);
-            if let Ok(export_state) = self.recv_export.recv() {
-                self.export_state = export_state;
+    }
+
+    pub fn exporting(&self) -> bool {
+        *self.export_state.lock() != ExportState::NotExporting
+    }
+
+    /// Schedule MIDI events and start to play music.
+    fn start_music(&mut self, state: &State) {
+        // Get the start time.
+        let start = state
+            .time
+            .ppq_to_samples(state.time.playback, self.framerate);
+
+        // Set the playback framerate.
+        let mut synth = self.synth.lock();
+        synth.set_sample_rate(self.framerate);
+
+        // Enqueue note events.
+        let mut midi_event_queue = self.midi_event_queue.lock();
+        for track in state.music.get_playable_tracks().iter() {
+            for note in track.get_playback_notes(state.time.playback) {
+                // Note-on event.
+                midi_event_queue.enqueue(
+                    state.time.ppq_to_samples(note.start, self.framerate),
+                    MidiEvent::NoteOn {
+                        channel: track.channel,
+                        key: note.note,
+                        vel: note.velocity,
+                    },
+                );
+                // Note-off event.
+                midi_event_queue.enqueue(
+                    state.time.ppq_to_samples(note.end, self.framerate),
+                    MidiEvent::NoteOff {
+                        channel: track.channel,
+                        key: note.note,
+                    },
+                );
+            }
+        }
+        // Sort the events by start time.
+        midi_event_queue.sort();
+
+        // Play music.
+        let mut play_state = self.play_state.lock();
+        *play_state = PlayState::Playing(start);
+    }
+
+    /// Stop ongoing music.
+    fn stop_music(&mut self, music: &Music) {
+        let mut synth = self.synth.lock();
+        for track in music.midi_tracks.iter() {
+            if synth
+                .send_event(MidiEvent::AllNotesOff {
+                    channel: track.channel,
+                })
+                .is_ok()
+                && synth
+                    .send_event(MidiEvent::AllSoundOff {
+                        channel: track.channel,
+                    })
+                    .is_ok()
+            {}
+        }
```

Comment:
Dropping a lock after using it as early as possible ensures that other threads need to block as little as possible.

```suggestion
        }
        drop(synth);
```